linker/src/lib.rs

```diff
@@ -82,27 +83,45 @@ pub fn link<T: FieldElement>(graph: PILGraph<T>) -> Result<PILFile<T>, Vec<Strin
 
                 let params = to.function.params;
 
-                let to_namespace = to.loc.clone().to_string();
+                let to_namespace = to.machine.location.clone().to_string();
 
                 let rhs = SelectedExpressions {
-                    selector: Some(namespaced_reference(to_namespace.clone(), to.latch)),
-                    expressions: once(namespaced_reference(to_namespace.clone(), to.function_id))
-                        .chain(
-                            params
-                                .inputs
-                                .params
-                                .iter()
-                                .chain(params.outputs.iter().flat_map(|o| o.params.iter()))
-                                .map(|i| {
-                                    namespaced_reference(to_namespace.clone(), i.name.clone())
-                                }),
-                        )
-                        .collect(),
+                    selector: Some(namespaced_reference(to_namespace.clone(), to.machine.latch)),
+                    expressions: once(namespaced_reference(
+                        to_namespace.clone(),
+                        to.machine.function_id,
+                    ))
+                    .chain(
+                        params
+                            .inputs
+                            .params
+                            .iter()
+                            .chain(params.outputs.iter().flat_map(|o| o.params.iter()))
+                            .map(|i| namespaced_reference(to_namespace.clone(), i.name.clone())),
+                    )
+                    .collect(),
                 };
 
                 let lookup = PilStatement::PlookupIdentity(0, lhs, rhs);
                 pil.push(lookup);
             }
+
+            if location == Location::default().join("main") {
+                assert!(
+                    graph.entry_points.len() <= 1,
+                    "Program must have at most 1 entry point, found {}",
+                    graph.entry_points.len()
+                );
+                if let Some(entry_point) = graph.entry_points.first() {
+                    let entry_point_id = entry_point.id;
+                    let function_id = main_machine.function_id.clone();
+                    // call the first function by initialising `function_id` to the first function
+                    pil.push(parse_pil_statement(&format!(
+                        "first_step * ({function_id} - {entry_point_id}) = 0"
```

Comment:
What I can do is redeclaring a new `first_step` for the linker step, so that it's all self contained. Then all the `first_step` can be optimised away.

Reply:
yea I don't have a solution for it myself, but self-contained sounds good. What I'd like to avoid is assuming that something called `first_step` exists here, although currently there are no connections to it's declaration. So if that changes wherever it is now, this breaks in a separate step

Reply:
Totally agree, self contained is the way then